## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :x: [more...](#c73ad0b5-ca38-47c2-a4de-4fcab10a7f58) |
| mint-compression-dist-xl.sh | :x: [more...](#e5d84619-9ec8-429e-b739-260974547f1d) |
| mint-compression-fs.sh | :x: [more...](#1cdd5fb0-915b-42c7-93e2-3030cf3adb05) |
| mint-fs.sh | :x: [more...](#083fc52c-7809-4ca2-b147-ce6da7b8d829) |
| mint-dist-xl.sh | :x: [more...](#0af1ce4f-78b2-4caf-a98b-a4fdb8486830) |
<a name="1cdd5fb0-915b-42c7-93e2-3030cf3adb05"></a>
### 8254-a5761ff/mint-compression-fs.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.59:30898
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp c786c5bef1d3:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 0 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 41 seconds
(4/14) Running aws-sdk-ruby tests ... done in 1 seconds
(5/14) Running awscli tests ... done in 52 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 26 seconds
(8/14) Running minio-dotnet tests ... done in 20 seconds
(9/14) Running minio-go tests ... done in 15 seconds
(10/14) Running minio-java tests ... done in 16 seconds
(11/14) Running minio-js tests ... FAILED in 16 seconds
{
  "name": "minio-js",
  "function": "fPutObject(bucketName, objectName, filePath, callback)",
  "args": "bucketName:minio-js-test-c5edcd25-029f-4045-a5e6-3e4fb3ef4992, objectName:datafile-65-MB, filePath:/tmp/datafile-65-MB",
  "duration": 3336,
  "status": "FAIL",
  "error": "Error: read ECONNRESET at exports._errnoException (util.js:1020:11) at TCP.onread (net.js:580:26)"
}

Executed 10 out of 14 tests successfully.
```
<a name="c73ad0b5-ca38-47c2-a4de-4fcab10a7f58"></a>
### 8254-a5761ff/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.60:31108
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp b92a3645f2cd:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 5 minutes and 5 seconds
(2/14) Running aws-sdk-java tests ... done in 2 seconds
(3/14) Running aws-sdk-php tests ... done in 55 seconds
(4/14) Running aws-sdk-ruby tests ... done in 15 seconds
(5/14) Running awscli tests ... done in 2 minutes and 10 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 1 minutes and 0 seconds
(8/14) Running minio-dotnet tests ... done in 1 minutes and 19 seconds
(9/14) Running minio-go tests ... FAILED in 3 minutes and 26 seconds
{
  "args": {},
  "duration": 64712,
  "error": "The specified key does not exist.",
  "function": "testStorageClassMetadataCopyObject()",
  "message": "Stat failed",
  "name": "minio-go: testStorageClassMetadataCopyObject",
  "status": "FAIL"
}

Executed 8 out of 14 tests successfully.
```
<a name="e5d84619-9ec8-429e-b739-260974547f1d"></a>
### 8254-a5761ff/mint-compression-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.53:31464
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 670b3215a83a:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 1 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 47 seconds
(4/14) Running aws-sdk-ruby tests ... done in 4 seconds
(5/14) Running awscli tests ... FAILED in 1 minutes and 57 seconds
{
  "name": "awscli",
  "duration": 63421,
  "function": "aws --endpoint-url http://72.28.97.53:31464 s3 sync --no-progress /mint/data s3://awscli-mint-test-bucket-14709/\n",
  "status": "FAIL",
  "error": "upload: ../../../data/datafile-0-b to s3://awscli-mint-test-bucket-14709/datafile-0-bupload: ../../../data/datafile-100-kB to s3://awscli-mint-test-bucket-14709/datafile-100-kBupload: ../../../data/datafile-1-kB to s3://awscli-mint-test-bucket-14709/datafile-1-kBupload: ../../../data/datafile-1-b to s3://awscli-mint-test-bucket-14709/datafile-1-bupload: ../../../data/datafile-1-MB to s3://awscli-mint-test-bucket-14709/datafile-1-MBupload: ../../../data/datafile-1.03-MB to s3://awscli-mint-test-bucket-14709/datafile-1.03-MBupload: ../../../data/datafile-33-kB to s3://awscli-mint-test-bucket-14709/datafile-33-kBupload: ../../../data/datafile-5-MB to s3://awscli-mint-test-bucket-14709/datafile-5-MBupload: ../../../data/datafile-5243880-b to s3://awscli-mint-test-bucket-14709/datafile-5243880-bupload: ../../../data/datafile-6-MB to s3://awscli-mint-test-bucket-14709/datafile-6-MBupload: ../../../data/datafile-10-kB to s3://awscli-mint-test-bucket-14709/datafile-10-kBupload: ../../../data/datafile-10-MB to s3://awscli-mint-test-bucket-14709/datafile-10-MBupload: ../../../data/datafile-65-MB to s3://awscli-mint-test-bucket-14709/datafile-65-MBupload failed: ../../../data/datafile-129-MB to s3://awscli-mint-test-bucket-14709/datafile-129-MB An error occurred (InvalidPart) when calling the CompleteMultipartUpload operation: One or more of the specified parts could not be found.  The part may not have been uploaded, or the specified entity tag may not match the part's entity tag.upload: ../../../data/datafile-11-MB to s3://awscli-mint-test-bucket-14709/datafile-11-MB"
}

Executed 4 out of 14 tests successfully.
```
<a name="083fc52c-7809-4ca2-b147-ce6da7b8d829"></a>
### 8254-a5761ff/mint-fs.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.54:30419
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 04fad34dd859:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 0 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 42 seconds
(4/14) Running aws-sdk-ruby tests ... done in 2 seconds
(5/14) Running awscli tests ... done in 1 minutes and 51 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 33 seconds
(8/14) Running minio-dotnet tests ... done in 26 seconds
(9/14) Running minio-go tests ... done in 42 seconds
(10/14) Running minio-java tests ... done in 45 seconds
(11/14) Running minio-js tests ... done in 43 seconds
(12/14) Running minio-py tests ... FAILED in 1 minutes and 56 seconds
{
  "duration": 57,
  "name": "minio-py:test_presigned_get_object_expiry_5sec",
  "function": "presigned_get_object(bucket_name, object_name, expires, response_headers, request_date)",
  "args": {
    "response._body": "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<Error><Code>AccessDenied</Code><Message>Request has expired</Message><Key>02521f9c-fbd9-441a-8e44-8214a10ad90b</Key><BucketName>minio-py-test-3a562915-7a47-411a-bd49-e341dc7deb7f</BucketName><Resource>/minio-py-test-3a562915-7a47-411a-bd49-e341dc7deb7f/02521f9c-fbd9-441a-8e44-8214a10ad90b</Resource><RequestId>15C52EF697F1EE75</RequestId><HostId>4b8c547d-1202-4c93-a889-fc7e1159a9ff</HostId></Error>",
    "bucket_name": "minio-py-test-3a562915-7a47-411a-bd49-e341dc7deb7f",
    "response.status": 403,
    "response.reason": "Forbidden",
    "response.headers": "{\"_container\": {\"accept-ranges\": [\"Accept-Ranges\", \"bytes\"], \"content-length\": [\"Content-Length\", \"440\"], \"content-security-policy\": [\"Content-Security-Policy\", \"block-all-mixed-content\"], \"content-type\": [\"Content-Type\", \"application/xml\"], \"server\": [\"Server\", \"MinIO/DEVELOPMENT.2019-09-17T08-54-07Z\"], \"vary\": [\"Vary\", \"Origin\"], \"x-amz-request-id\": [\"X-Amz-Request-Id\", \"15C52EF697F1EE75\"], \"x-xss-protection\": [\"X-Xss-Protection\", \"1; mode=block\"], \"date\": [\"Date\", \"Tue, 17 Sep 2019 09:14:41 GMT\"]}}",
    "presigned_get_object_url": "http://72.28.97.54:30419/minio-py-test-3a562915-7a47-411a-bd49-e341dc7deb7f/02521f9c-fbd9-441a-8e44-8214a10ad90b?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=minio%2F20190917%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20190917T091436Z&X-Amz-Expires=5&X-Amz-SignedHeaders=host&X-Amz-Signature=0870fcc62679fabb93e6a736c51c9236575d4e3ec19273be6b8fcfea6efcd71f",
    "object_name": "02521f9c-fbd9-441a-8e44-8214a10ad90b"
  },
  "message": "AccessDenied: message: Access Denied",
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 1341, in test_presigned_get_object_expiry_5sec\n    object_name).get_exception()\nminio.error.AccessDenied: AccessDenied: message: Access Denied\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 2036, in main\n    test_presigned_get_object_expiry_5sec(client, log_output)\n  File \"/mint/run/core/minio-py/tests.py\", line 1355, in test_presigned_get_object_expiry_5sec\n    raise Exception(err)\nException: AccessDenied: message: Access Denied\n",
  "status": "FAIL"
}

Executed 11 out of 14 tests successfully.
```
<a name="0af1ce4f-78b2-4caf-a98b-a4fdb8486830"></a>
### 8254-a5761ff/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.54:30697
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 931ad66bae65:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 3 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 44 seconds
(4/14) Running aws-sdk-ruby tests ... done in 5 seconds
(5/14) Running awscli tests ... done in 2 minutes and 4 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 46 seconds
(8/14) Running minio-dotnet tests ... done in 52 seconds
(9/14) Running minio-go tests ... done in 1 minutes and 21 seconds
(10/14) Running minio-java tests ... done in 47 seconds
(11/14) Running minio-js tests ... done in 55 seconds
(12/14) Running minio-py tests ... FAILED in 1 minutes and 46 seconds
{
  "name": "minio-py:test_list_objects_with_1001_files",
  "duration": 80106,
  "function": "list_objects(bucket_name, prefix, recursive)",
  "status": "FAIL",
  "args": {
    "bucket_name": "minio-py-test-790c7f27-94fd-4ebe-a935-d7608412804b",
    "object_name": "d91ffce3-7cc6-4c8f-90df-28ddfd168223_0 ~ d91ffce3-7cc6-4c8f-90df-28ddfd168223_1000"
  },
  "message": "BucketNotEmpty: message: The bucket you tried to delete is not empty.",
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 1168, in test_list_objects_with_1001_files\n    client.remove_bucket(bucket_name)\n  File \"/usr/local/lib/python3.5/dist-packages/minio/api.py\", line 417, in remove_bucket\n    self._url_open('DELETE', bucket_name=bucket_name)\n  File \"/usr/local/lib/python3.5/dist-packages/minio/api.py\", line 1952, in _url_open\n    object_name).get_exception()\nminio.error.BucketNotEmpty: BucketNotEmpty: message: The bucket you tried to delete is not empty.\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 2024, in main\n    test_list_objects_with_1001_files(client, log_output)\n  File \"/mint/run/core/minio-py/tests.py\", line 1170, in test_list_objects_with_1001_files\n    raise Exception(err)\nException: BucketNotEmpty: message: The bucket you tried to delete is not empty.\n"
}

Executed 11 out of 14 tests successfully.
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -56,6 +56,3 @@ before_script:
 
 before_install:
  - if [[ "$TRAVIS_OS_NAME" == "linux" ]]; then nvm install 11.10.1 ; fi
-
-after_success:
-- bash <(curl -s https://codecov.io/bash)

--- a/cmd/gateway/hdfs/gateway-hdfs.go
+++ b/cmd/gateway/hdfs/gateway-hdfs.go
@@ -507,12 +507,34 @@ func (n *hdfsObjects) GetObject(ctx context.Context, bucket, key string, startOf
 	return hdfsToObjectErr(ctx, err, bucket, key)
 }
 
+func (n *hdfsObjects) isObjectDir(ctx context.Context, bucket, object string) bool {
+	f, err := n.clnt.Open(minio.PathJoin(hdfsSeparator, bucket, object))
+	if err != nil {
+		if os.IsNotExist(err) {
+			return false
+		}
+		logger.LogIf(ctx, err)
+		return false
+	}
+	defer f.Close()
+	fis, err := f.Readdir(1)
+	if err != nil {
+		logger.LogIf(ctx, err)
+		return false
+	}
+	return len(fis) == 0
+}
+
 // GetObjectInfo reads object info and replies back ObjectInfo.
 func (n *hdfsObjects) GetObjectInfo(ctx context.Context, bucket, object string, opts minio.ObjectOptions) (objInfo minio.ObjectInfo, err error) {
 	_, err = n.clnt.Stat(minio.PathJoin(hdfsSeparator, bucket))
 	if err != nil {
 		return objInfo, hdfsToObjectErr(ctx, err, bucket)
 	}
+	if strings.HasSuffix(object, hdfsSeparator) && !n.isObjectDir(ctx, bucket, object) {
+		return objInfo, hdfsToObjectErr(ctx, os.ErrNotExist, bucket, object)
+	}
+
 	fi, err := n.clnt.Stat(minio.PathJoin(hdfsSeparator, bucket, object))
 	if err != nil {
 		return objInfo, hdfsToObjectErr(ctx, err, bucket, object)
@@ -536,7 +558,7 @@ func (n *hdfsObjects) PutObject(ctx context.Context, bucket string, object strin
 	name := minio.PathJoin(hdfsSeparator, bucket, object)
 
 	// If its a directory create a prefix {
-	if strings.HasSuffix(object, hdfsSeparator) {
+	if strings.HasSuffix(object, hdfsSeparator) && r.Size() == 0 {
 		if err = n.clnt.MkdirAll(name, os.FileMode(0755)); err != nil {
 			n.deleteObject(minio.PathJoin(hdfsSeparator, bucket), name)
 			return objInfo, hdfsToObjectErr(ctx, err, bucket, object)

